llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -3472,6 +3472,20 @@ InstructionCost AArch64TTIImpl::getArithmeticInstrCost(
           Cost *= 4;
         return Cost;
       } else {
+        // If the information about individual scalars being vectorized is
+        // available, this yeilds better cost estimation.
+        if (auto *VTy = dyn_cast<FixedVectorType>(Ty); VTy && !Args.empty()) {
+          InstructionCost InsertExtractCost =
+              ST->getVectorInsertExtractBaseCost();
+          Cost = (3 * InsertExtractCost) * VTy->getNumElements();
+          for (int i = 0, Sz = Args.size(); i < Sz; i += 2) {
+            Cost += getArithmeticInstrCost(
+                Opcode, VTy->getScalarType(), CostKind,
+                TTI::getOperandInfo(Args[i]), TTI::getOperandInfo(Args[i + 1]));
+          }
+          return Cost;
+        }
```

Comment:
> I don't understand what are you trying to do here. Why are you adding the cost of the scalar ops here? They are already counted in the cost model

AArch64 Neon does not support vector division natively. It scalarizes the vector division as can be seen [here](https://godbolt.org/z/EaxKsGe3b). 

Now coming to the cost modelling part: This code section is calculating the cost of scalarizing the vector division. Thus,
```
vector cost = 
for each lane(cost of extract for both div operands) + 
for each lane(cost of scalar division) + 
for each lane(cost of insert into the result vector) 
```
This vector cost was being over-calculated thus preferring scalar code. This patch just tries to rectify this.

Reply:
Then, you should not build the vector node at all. I would suggest instead that you check how the `getScalarsVectorizationState` function in SLP works, add a check for NEON division (maybe add a new entry in TTI to check if the vector operation is legal and won't be scalarized), and return that for NEON, it should build the `TreeEntry::NeedToGather` node.

Reply:
> Then, you should not build the vector node at all. I would suggest instead that you check how the `getScalarsVectorizationState` function in SLP works, add a check for NEON division (maybe add a new entry in TTI to check if the vector operation is legal and won't be scalarized), and return that for NEON, it should build the `TreeEntry::NeedToGather` node.

Two issues which I can think of with this approach:
1. The operands of the div operation wont get vectorized.
2. Lot of checks are in place that alter the div cost based on information about  its operands(e.g. if operands are uniform/constant/pow-of-2 etc.). Hence, factoring out code just to check if the div operation would be scalarized is difficult and would result in duplication of code.

Reply:
> 1. That's fine. If you think they should be vectorized, a new vectorization strategy should be implemented instead.

Don't you think lowering the division vector in some late IR pass is easy than stopping vectorization at div ? One more advantage of lowering later is lowering is transparent to the user and done at single place rather than at multiple places.
This type of vector lowering is already being done at other places e.g. masked_load

> 2\. No, it is just a simple legality check. If the instruction should be scalarized, it should be built into a build vector node; it should not be vectorized.

Just need a clarification: my understanding of `TreeEntry::NeedToGather` is you build a vector out of scalars. Is that correct?
Now, coming to the legality checks. Yes, in the end, its legality check but I need to check all the  contraints, which are currently in `getArithmeticInstrCost` once again. And I am trying to avoid that route just for reasons mentioned in my comment on (1).

Reply:
> > 1. That's fine. If you think they should be vectorized, a new vectorization strategy should be implemented instead.
> 
> Don't you think lowering the division vector in some late IR pass is easy than stopping vectorization at div ? One more advantage of lowering later is lowering is transparent to the user and done at single place rather than at multiple places. This type of vector lowering is already being done at other places e.g. masked_load
> 

It just breaks the vectorizer design. If the node should be scalarized, the vectorization model should know about it and scalarize it. Otherwise, it complicates the whole vectorization process and makes it miss some possible vectorization opportunities. I assume, you missed the cost of extraction of the operands in your cost model changes?

> > 2. No, it is just a simple legality check. If the instruction should be scalarized, it should be built into a build vector node; it should not be vectorized.
> 
> Just need a clarification: my understanding of `TreeEntry::NeedToGather` is you build a vector out of scalars. Is that correct? Now, coming to the legality checks. Yes, in the end, its legality check but I need to check all the contraints, which are currently in `getArithmeticInstrCost` once again. And I am trying to avoid that route just for reasons mentioned in my comment on (1).

You can check these constraints before and command vectorizer to build a NeedToGather node. You should not use getArithmeticInstrCost in this case, some other function should check if it is legal to vectorize these scalars here

Reply:
> Otherwise, it complicates the whole vectorization process and makes it miss some possible vectorization opportunities. I assume, you missed the cost of extraction of the operands in your cost model changes?

No, in fact the patch is accurately calculating the cost of:
```
for every lane: 2 extracts and 1 div and 1 insert
See
   Cost = (3 * InsertExtractCost) * VTy->getNumElements();   // Insert/Extract cost is same
```
So, no way it is missing out on opportunities and by no means its complicating the process.